{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "27dccb24-2fef-4b5d-832c-f161076d8b2f",
   "metadata": {},
   "source": [
    "## Customized Grouping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "022e072f-aeb8-4ed2-91ff-b5b267da8729",
   "metadata": {},
   "source": [
    "Customized Grouping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc7ac35-3441-4470-b118-aa2a98ce3742",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c7b25a-e90f-461a-ad96-f9a8152f3bf8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5819f1d4-0364-41c1-9392-9c40c01f19b3",
   "metadata": {},
   "source": [
    "a, the taxonomic grouping microbes by truncating taxonomic levels in the phylogenetic tree (the 1, 3, and 10 clusters are generated by truncating the Control, Kingdom, and Phylum levels, respectively. c is the number of the channels, e. g., c=10 means that the number of channels of the 2D-microbiomeprint is 10). \n",
    "b, the metagenomic grouping by microbes specifying the number of clusters in the hierarchical clustering tree. To make a fair comparison with the taxonomic-based grouping, the same number of the clusters are specified in metagenomic-based grouping (i. e., c=1, 3, 10, respectively.). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a2a16c-f1a5-4556-a009-bd14d67c54be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
